### PR TITLE
Mixed precision fusing

### PR DESCRIPTION
**Pull request by nluehr**
Independent tensor fusion by dtype to enable more fusion in mixed precision cases.

### CODE REVIEW

**Comment by alsrgv** [comment]
@nluehr, thanks for the PR!  Do you have a benchmark showing the impact?

I'd also be curious to see if this introduces too much look-ahead since tensors that are "ready-to-reduce" are not necessarily fully computed on GPU, hence there's waiting that happens at the beginning of `PerformOperation()`.

One of the possible mitigations is to [wait for tensors](https://github.com/uber/horovod/commit/47bec61fe8f8a06d9230a6c1a55a447408e53f41) before considering them "ready-to-reduce", but that change was not universally better and so is not committed yet.  Could you see if it helps in your use case?

**Comment by alsrgv** [comment]
@nluehr, any luck with the benchmark?

**Comment by romerojosh** [comment]
Hey Alex, 
Sorry this took a while but we finally have some benchmark results for this PR. We ran RN50 with TensorFlow on a multi-node DGX1 cluster with mixed-precision allreduces (FP32 for batch norms, FP16 for everything else). We tested on 1 to 8 nodes (8 - 64 V100 GPUs) both a baseline case without the PR, a case with the PR only, and a case with PR plus the wait for ready mitigation you suggested with per GPU batchsizes of 32, 64, and 128.  Essentially, the finding is that for multi-node cases where internode communication is involved, the additional fusing of tensors enabled by this PR provides a good performance boost; however, for single node, performance is degraded as NCCL seems to be capable of better handling the many small allreduces generated in this configuration. The wait for ready mitigation seems to hurt single node perf relative to the baseline a bit more, but has only a slight (negative) impact on the multi-node cases when combined with the PR. See the benchmark results below:

```Case: TF + Horovod 15.2 (baseline)
BS: 32
Nodes: 1, Samples:  4431.82 4441.99 4390.04 4437 Average: 4425.21
Nodes: 2, Samples:  8219.8 8261.67 8203.57 8220.77 Average: 8226.45
Nodes: 4, Samples:  14885 14653.8 14894.4 14498.6 Average: 14732.9
Nodes: 8, Samples:  22088.9 21755.6 21418.7 20082 Average: 21336.3
BS: 64
Nodes: 1, Samples:  5549.05 5545.56 5547.7 Average: 5547.44
Nodes: 2, Samples:  10550.5 10402.2 10473.9 10571.8 Average: 10499.6
Nodes: 4, Samples:  20319.8 20345.7 20331.5 20377.3 Average: 20343.6
Nodes: 8, Samples:  35539.6 34260.2 35993.4 35982.6 Average: 35443.9
BS: 128
Nodes: 1, Samples:  6452.24 6489.38 6476.72 6493.14 Average: 6477.87
Nodes: 2, Samples:  12762.6 12630.8 12672.8 Average: 12688.7
Nodes: 4, Samples:  24944.5 24827.2 25011.3 24838.6 Average: 24905.4
Nodes: 8, Samples:  45645.6 44465.7 45513.3 45691.3 Average: 45329

Case: TF + Horovod 15.2 + mixed precision fusion PR
BS: 32
Nodes: 1, Samples:  4474.95 4507.17 4477.17 4506.36 Average: 4491.41
Nodes: 2, Samples:  8686.57 8712.69 8752.92 8760.13 Average: 8728.08
Nodes: 4, Samples:  17169.7 17184.9 16800.3 16923.5 Average: 17019.6
Nodes: 8, Samples:  24348.9 26468.7 28805.9 26572 Average: 26548.9
BS: 64
Nodes: 1, Samples:  5462.62 5470.42 5462.33 5458.07 Average: 5463.36
Nodes: 2, Samples:  10780.5 10785.5 10793.8 10868.3 Average: 10807
Nodes: 4, Samples:  21354 21350.7 21380.4 21414.9 Average: 21375
Nodes: 8, Samples:  39865.4 40936.6 41763 41204 Average: 40942.2
BS: 128
Nodes: 1, Samples:  6391.85 6392.59 6370.56 6391.42 Average: 6386.6
Nodes: 2, Samples:  12637 12709.1 12685.1 12721.5 Average: 12688.2
Nodes: 4, Samples:  25290.4 25306.3 25254.7 25276.3 Average: 25281.9
Nodes: 8, Samples:  50085.7 49331.7 50019.2 49764.5 Average: 49800.3

Case: TF + Horovod 15.2 + mixed precision fusion PR + wait for ready mitigation
BS: 32
Nodes: 1, Samples:  4203.85 4207.05 4194.44 4204.2 Average: 4202.39
Nodes: 2, Samples:  8375.94 8363.53 8352.87 8347.32 Average: 8359.92
Nodes: 4, Samples:  16465.5 16437.2 16406 16260.4 Average: 16392.3
Nodes: 8, Samples:  30993.6 31113.8 30861.1 30948.7 Average: 30979.3
BS: 64
Nodes: 1, Samples:  5310.79 5311.89 5303.67 5299.44 Average: 5306.45
Nodes: 2, Samples:  10584.4 10584.6 10586.9 10552.6 Average: 10577.1
Nodes: 4, Samples:  20651.7 20807.7 20693.9 20603.9 Average: 20689.3
Nodes: 8, Samples:  39715.9 40159.1 40106.2 39990.7 Average: 39993
BS: 128
Nodes: 1, Samples:  6313.15 6310.99 6299.21 6310.28 Average: 6308.41
Nodes: 2, Samples:  12560.4 12581.7 12579.3 12560.6 Average: 12570.5
Nodes: 4, Samples:  24876.6 24901.9 24921.1 24742.8 Average: 24860.6
Nodes: 8, Samples:  48874.6 48685.5 48598.8 48938.5 Average: 48774.4```

Let us know how you'd like to proceed when you have a chance.

**Comment by alsrgv** [comment]
@romerojosh, sorry for the delay.  I'm wondering if you could try to adjust `HOROVOD_FUSION_THRESHOLD` in your single-node tests and see if detrimental effect of fusion on single-node performance could be alleviated?  If that's the case, `HOROVOD_AUTOTUNE=1` should be able to find high-performant combination automatically.

**Comment by romerojosh** [comment]
@alsrgv, I generated some additional single node data with fusion thresholds of 0, 2, 4, and 8 MiB. The general finding is that modifying `HOROVOD_FUSION_THRESHOLD` does not look like it will recover the lost performance. It looks like from no fusion up to 8 MiB, the trend is that performance increases and roughly approaches at the single node performance posted previously. Also important to note is that without the wait for ready mitigation, the no fusion (threshold = 0) case with the PR seems to suffer a large performance drop, potentially a function of more waiting for data that occurs in this configuration. The wait for ready mitigation gets rid of this drop, but limits any performance gains from increased fusion.   

Here is the actual data:
```
Case: TF + Horovod 15.2 (baseline)
BS: 32
Threshold: 0, Samples:  4449.5 Average: 4449.5
Threshold: 2097152, Samples:  4460.16 Average: 4460.16
Threshold: 4194304, Samples:  4372.02 Average: 4372.02
Threshold: 8388608, Samples:  4438.32 Average: 4438.32
BS: 64
Threshold: 0, Samples:  5506.25 Average: 5506.25
Threshold: 2097152, Samples:  5564.83 Average: 5564.83
Threshold: 4194304, Samples:  5522.14 Average: 5522.14
Threshold: 8388608, Samples:  5458.06 Average: 5458.06
BS: 128
Threshold: 0, Samples:  6470.1 Average: 6470.1
Threshold: 2097152, Samples:  6488.94 Average: 6488.94
Threshold: 4194304, Samples:  6481.04 Average: 6481.04
Threshold: 8388608, Samples:  6450.92 Average: 6450.92

Case: TF + Horovod 15.2 + mixed precision fusion PR
BS: 32
Threshold: 0, Samples:  4139.62 4129.99 Average: 4134.81
Threshold: 2097152, Samples:  4376.55 Average: 4376.55
Threshold: 4194304, Samples:  4424.12 Average: 4424.12
Threshold: 8388608, Samples:  4489.92 Average: 4489.92
BS: 64
Threshold: 0, Samples:  4973.76 5013.51 Average: 4993.64
Threshold: 2097152, Samples:  5424.4 Average: 5424.4
Threshold: 4194304, Samples:  5437.65 Average: 5437.65
Threshold: 8388608, Samples:  5406.6 Average: 5406.6
BS: 128
Threshold: 0, Samples:  6003.49 6018.1 Average: 6010.8
Threshold: 2097152, Samples:  6354.89 Average: 6354.89
Threshold: 4194304, Samples:  6360.73 Average: 6360.73
Threshold: 8388608, Samples:  6351.15 Average: 6351.15

Case: TF + Horovod 15.2 + mixed precision fusion PR + wait for ready mitigation
BS: 32
Threshold: 0, Samples:  4219.23 Average: 4219.23
Threshold: 2097152, Samples:  4195.69 Average: 4195.69
Threshold: 4194304, Samples:  4194.19 Average: 4194.19
Threshold: 8388608, Samples:  4199.96 Average: 4199.96
BS: 64
Threshold: 0, Samples:  5325.94 Average: 5325.94
Threshold: 2097152, Samples:  5304.27 Average: 5304.27
Threshold: 4194304, Samples:  5313.65 Average: 5313.65
Threshold: 8388608, Samples:  5304.15 Average: 5304.15
BS: 128
Threshold: 0, Samples:  6300.94 Average: 6300.94
Threshold: 2097152, Samples:  6308.67 Average: 6308.67
Threshold: 4194304, Samples:  6302.06 Average: 6302.06
Threshold: 8388608, Samples:  6299.58 Average: 6299.58
```

While single node performance is impacted by this change, I think the multinode performance improvements make this PR worthwhile. Perhaps making this additional fusion available via a runtime flag would be a feasible way forward?

**Comment by yuxihu** [comment]
I tried this PR with MXNet. It can boost the training throughput by about 16% on 64 V100 GPUs. I used 256 per GPU batch size. It is significant performance improvement for multi-node training.

**Comment by alsrgv** [comment]
@romerojosh, got it, thanks for running more tests!  Could you rebase it w.r.t. latest changes in master?  I'll try it out in my environment as well.

**Comment by romerojosh** [comment]
@alsrgv PR has been rebased to current master but CI still seems to fail due to issues unrelated to these changes (network timeouts). Let us know how it works out for you in your environment and if you have any suggested changes.

File: horovod/common/operations.cc [hunk 1]
```diff
@@ -2135,94 +2136,100 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
+          responses.pop_front();
+          int64_t tensor_size = 0;
+
+          if (response.response_type() == MPIResponse::ResponseType::ALLREDUCE) {
+            // Attempt to add more responses to this fused response.
+            auto& entry = state.tensor_table[response.tensor_names()[0]];
+            tensor_size = entry.tensor->size();
+
+            while (!responses.empty()) {
+              auto new_response = responses.front();
```
Comment (shahzadlone) [inline horovod/common/operations.cc]: Perhaps ```new_reponse``` can be ```const``` if it isn't muted / modified ?

File: horovod/common/operations.cc [hunk 2]
```diff
@@ -2135,94 +2136,100 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
+                  tensor_size + new_tensor_size <= TensorFusionThresholdBytes()) {
+                tensor_size += new_tensor_size;
+                response.add_tensor_name(new_response.tensor_names()[0]);
+                responses.pop_front();
+              } else {
+                // Don't try to fuse additional tensors since they are usually
+                // computed in order of requests and skipping tensors may mean
+                // that the batch will have to wait longer while skipped tensors
+                // could be reduced at that time.
+                break;
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/common/operations.cc]: I did know this would lead to some misordering, as the map data structure will always force all FP16 ops to take precedence over FP32 operations, which in general might not be a nice thing to do. Your suggested version would definitely fix that. As a side effect, this change would introduce look ahead across response type boundaries though (I don't think this is necessarily a bad thing, but if you think the break condition should also maintain this behavior, let me know).

**Comment by romerojosh** [comment]
You are probably right about your benchmark only using FP16 with no FP32. The benchmark I was using for these results is the RN50 example in the NVIDIA TensorFlow 19.01 container (available at https://ngc.nvidia.com/catalog/containers/nvidia:tensorflow), with modifications to the included `nvidia-examples/cnn/nvuntils/runner.py` script to allow FP16 allreduces (except for batch norm which is kept in FP32). Those changes are in the attached [runner_patch.txt].(https://github.com/uber/horovod/files/2824724/runner_patch.txt)

I run the RN50 command to run for 3 epochs:
`mpirun -np <N> --allow-run-as-root python /workspace/nvidia-examples/cnn/resnet.py --layers 50 --data_dir <Path to TFRecord dataset> --display_every 10 --num_iter 3 --iter_unit epoch`

Let me know if you need any more info on the benchmark or just want copies of the scripts to run in your own environment.

As a side note, this configuration mimics what MXNet does in its mixed precision mode (FP32 batchnorm, FP16 everything else).

File: horovod/common/operations.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (alsrgv) [reply horovod/common/operations.cc]: I think we could introduce look-ahead penalty expressed in terms of size, and approximate the cost of obtaining the output by its size.  Then, we could break once the size of all skipped tensors is greater than X.  I don't know the good value for X, maybe 50% of fusion buffer size?  Or 10%... or 100%...  Worth exploring, and perhaps expose another tuning parameter.

[the remark this answers is not in the document]
Comment (alsrgv) [reply horovod/common/operations.cc]: Or possibly, we could continue while `sum(existing output size) + sum(skipped size) <= TensorFusionThresholdBytes()`.  That would avoid another hyperparameter and still feels good.

[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/common/operations.cc]: I agree on limiting based on sum of skipped messages (much better than adding another tunable parameter too). I went ahead a pushed a change to do exactly this so please take a look.

It should have mostly no impact on my current benchmark case since this logic should result in the same fusing (total gradient size per step is under the default 64 MiB fusion threshold so each cycle should launch two fully fused allreduces, one FP16 and one FP32). The order that these two operations occur might change, as it now will honor whatever response was first in queue. 

The more preserved ordering may improve the single node cases with modified `HOROVOD_FUSION_THRESHOLD` so I will rerun those tests to see.

**Comment by alsrgv** [comment]
Awesome!  Final nit - can we apply the same logic to Allgather, too, to keep them aligned?

**Comment by romerojosh** [comment]
No problem! Just added the logic to the fused Allgather path now.
